### PR TITLE
Fix TidbMonitor several error

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb-operator/issues/1949

### What is changed and how does it work?

Fix following problem:

1. Add check to guaranteeing the nodePort won't be changed if the service Type of tidbmonitor is Nodeport
2. Add EnvVar sort to avoid the monitor controller render different result from the same tidbmonitor spec.
3. Add LB IP configuration rendering.
4. Change the reloader servicename to keep as same as the helm template.

Related changes

 - Need to cherry-pick to the release branch

### Does this PR introduce a user-facing change?:

```release-note
1. Add check to guaranteeing the nodePort won't be changed if the serviceType of TidbMonitor is Nodeport
2. Add EnvVar sort to avoid the monitor controller render different result from the same TidbMonitor spec.
3. Fix the problem that Tidbmonitor LoadBalance IP is not used.
```

### CODE REVIEW

**Comment by Yisaer** [comment]
> please add some unit tests, especially UTs that will fail previously and could pass with these fixes

In this case, e2e test is more suitable and I add e2e case instead.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -767,6 +767,58 @@ var _ = ginkgo.Describe("[tidb-operator] TiDBCluster", func() {
...
+			}
+			if prometheusSvc.Spec.Ports[0].Name != "any-other-word" {
+				return false, nil
+			}
+			if prometheusSvc.Spec.Ports[0].NodePort != targetPort {
+				return false, nil
+			}
+			return true, nil
+		})
+		framework.ExpectNoError(err, "second update tidbmonitor service error")
```
Comment (Yisaer) [inline tests/e2e/tidbcluster/tidbcluster.go]: I updated the port-name here, and check whether the nodePort is unchanged in this e2e case.

File: pkg/controller/generic_control.go [hunk 2]
```diff
@@ -271,8 +271,27 @@ func (w *typedWrapper) CreateOrUpdateService(controller runtime.Object, svc *cor
...
+			// If the existed service and the desired service is NodePort or LoadBalancerType, we should keep the nodePort unchanged.
+			if (serviceType == corev1.ServiceTypeNodePort || serviceType == corev1.ServiceTypeLoadBalancer) &&
+				(desiredSvc.Spec.Type == corev1.ServiceTypeNodePort || desiredSvc.Spec.Type == corev1.ServiceTypeLoadBalancer) {
+				for i, dport := range existingSvc.Spec.Ports {
+					for _, eport := range ports {
+						// Because the portName could be edited,
+						// we use Port number to link the desired Service Port and the existed Service Port in the nested loop
+						if dport.Port == eport.Port {
+							dport.NodePort = eport.NodePort
+							existingSvc.Spec.Ports[i] = dport
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/controller/generic_control.go]: I don't think so. In this nested loop, the existingSvc's Spec here has become desired ( which is unchanged in this process) yet, and the ports represents the original existeingSvc's ports here. If the existed service have 3 ports and the desired service have ports, the service would finally have 2 ports.

File: pkg/monitor/monitor/monitor_manager.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ import (
 type MonitorManager struct {
 	typedControl     controller.TypedControlInterface
 	deploymentLister appslisters.DeploymentLister
+	svcLister        corelisters.ServiceLister
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: removed

File: pkg/controller/generic_control.go [hunk 4]
```diff
@@ -271,8 +271,27 @@ func (w *typedWrapper) CreateOrUpdateService(controller runtime.Object, svc *cor
...
 			existingSvc.Spec.ClusterIP = clusterIp
+
+			// If the existed service and the desired service is NodePort or LoadBalancerType, we should keep the nodePort unchanged.
+			if (serviceType == corev1.ServiceTypeNodePort || serviceType == corev1.ServiceTypeLoadBalancer) &&
+				(desiredSvc.Spec.Type == corev1.ServiceTypeNodePort || desiredSvc.Spec.Type == corev1.ServiceTypeLoadBalancer) {
+				for i, dport := range existingSvc.Spec.Ports {
+					for _, eport := range ports {
+						// Because the portName could be edited,
+						// we use Port number to link the desired Service Port and the existed Service Port in the nested loop
+						if dport.Port == eport.Port {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/controller/generic_control.go]: updated.

File: pkg/controller/generic_control.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/controller/generic_control.go]: ok

**Pull request merged**